#80 - Initial support for auth-providers

javierarrieta opened the pull request:
This is an initial approach to work on auth-providers, still need to refresh tokens

doriordan commented on client/src/main/scala/skuber/api/Configuration.scala:
```diff
@@ -143,23 +143,59 @@ object Configuration {
-          password = optionalValueAt(userConfig, "password"))
+          }
+
+          val maybeAtuh = optionalValueAt[YamlMap](userConfig, "auth-provider") match {
```
I assume  'maybeAtuh' is a typo, should be 'maybeAuth'?

javierarrieta replied:
Yeah, thanks. Damn phat phingers

doriordan commented on client/src/main/scala/skuber/api/Configuration.scala:
```diff
@@ -143,23 +143,59 @@ object Configuration {
+              val userName = optionalValueAt(userConfig, "username")
+              val password = optionalValueAt(userConfig, "password")
+
+              (
```
I think a pattern match approach would be clearer, something like
```
  (userName,password,token,jwt,clientCertificate,clientKey) match {
    case (Some(u), Some(p),_,_,_,_)  => BasicAuth(u,p)
    ...
  }
```

doriordan commented on client/src/main/scala/skuber/api/package.scala:
```diff
@@ -51,11 +49,66 @@ package object client {
    )
 
-   case class AuthInfo(
+   sealed trait AuthInfo
```
I like this refactoring, the single AuthInfo class was getting cumbersome

doriordan commented on client/src/main/scala/skuber/api/package.scala:
```diff
@@ -51,11 +49,66 @@ package object client {
+       """GcpAuth(accessToken=<redacted>)""".stripMargin
+   }
+
+   case class AuthInfo1(
```
I assume `AuthInfo1` should be removed?

javierarrieta replied:
Yeah, I kept that to have easy the old code, should have removed before committing

javierarrieta commented:
Hi @doriordan , changed the requested code, thanks for the review. I am unsure if it is worth merging now or wait until I implement the token refresh as it might change the API a bit? Not sure when you want to release

doriordan commented:
Hi @javierarrieta if you are planning to add token refresh in the near future then I think it is best to hold off on the merge

javierarrieta commented:
Cool, will do that now when I have a chance, wanted to be sure you were happy with the approach first

Thanks!

javierarrieta commented:
Added refreshing of the Gcp token, not sure how oicd is refreshed. Refresh is done on demand with a hardcoded threshold of 30 seconds.

For a ouaht2 lib I implemented a state machine that will refresh the tokens, but might be too complex for this use case and also would need to change the api to `Future[String]`

Opinions @doriordan ?

javierarrieta commented on client/src/main/scala/skuber/api/package.scala:
```diff
@@ -51,37 +52,115 @@ package object client {
+  private[client] object GcpRefresh {
+    implicit val gcpRefreshReads: Reads[GcpRefresh] = (
+      (JsPath \ "credential" \ "access_token").read[String] and
+      (JsPath \ "credential" \ "token_expiry").read[Instant]
```
This should not be hardcoded, but don't know how to use JsPath to apply what there is in the config, I should evaluate the JsPath in these

```
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
```

javierarrieta commented:
Hi @doriordan any stopper for this? I could remove the hardcode and oidc refresh later if you are happy with the approach

doriordan commented:
Hi @javierarrieta I just want to have another look before I merge but I don't think there is any stopper for this

javierarrieta commented:
Cool, thanks.

nemo83 commented:
@javierarrieta @doriordan can this be used to authenticate with EKS?

javierarrieta commented:
Hi @nemo83 , haven't checked, hasn't been following this too much, any findings? Any way I can help?

nemo83 commented:
Compared to the java client and it doesn't seem that EKS is supported in this lib. Because of the very little code I had to write I decided to go ahead with the java lib instead. Thanks for replying to this.

javierarrieta commented:
Sure, no worries, sorry for the very late reply. Any link to the java lib so maybe worth contributing to this repo?

nemo83 commented:
This is the k8s java client thread for the EKS auth impl:

https://github.com/kubernetes-client/java/issues/315

doriordan commented:
Hi @javierarrieta sorry for late reply - thanks for the pointer and I'll take a look into what would be required to support EKS

javierarrieta commented:
Is this related to https://github.com/doriordan/skuber/issues/173 ?

doriordan commented:
@javierarrieta yes, need to revisit that and see if it is best to take that forward or a different take on it
